pallets/settlement/src/lib.rs

```diff
@@ -1118,22 +1269,119 @@ impl<T: Trait> Module<T> {
                 }
             }
 
+            enum FailureReason {
+                BaseTransfer,
+                ConfidentialTransfer,
+                UndefinedLeg,
+                InsufficientMercatAuthorizations,
+                GetMercatAccount,
+                GetMercatMediatorData,
+            }
+
             if !failed {
                 match with_transaction(|| {
                     for (leg_id, leg_details) in legs.iter().filter(|(leg_id, _)| {
                         let status = Self::instruction_leg_status(instruction_id, leg_id);
-                        status == LegStatus::ExecutionPending
+                        matches!(status, LegStatus::ExecutionPending)
                     }) {
-                        let result = T::Asset::base_transfer(
-                            leg_details.from,
-                            leg_details.to,
-                            &leg_details.asset,
-                            leg_details.amount,
-                        );
-                        if let Ok(post_info) = result {
-                            transaction_weight += post_info.actual_weight.unwrap_or_default();
-                        } else {
-                            return Err(leg_id);
+                        match leg_details {
+                            Leg::NonConfidentialLeg(leg) => {
+                                let result = T::Asset::base_transfer(
+                                    leg.from, leg.to, &leg.asset, leg.amount,
+                                );
+                                if let Ok(post_info) = result {
+                                    transaction_weight +=
+                                        post_info.actual_weight.unwrap_or_default();
+                                } else {
+                                    return Err((leg_id, FailureReason::BaseTransfer));
+                                }
+                            }
+                            Leg::ConfidentialLeg(leg) => {
```

Comment:
Splitting code executed in this match arm into a function of its own will make the code more readable imo.

Reply:
Yep, I suggested it in one of my previous comments. Satyam may have created a story for it.

Reply:
Here is the story that @satyamakgec created: https://polymath.atlassian.net/browse/MESH-1329